egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py

```diff
@@ -1039,3 +1040,1052 @@ def generate_pgru_config(self):
+            else:
+                raise Exception("In {0} of type {1}, unknown auxiliary output name {1}".format(self.layer_type, auxiliary_output))
+
+        return self.config['recurrent-projection-dim'] + self.config['non-recurrent-projection-dim']
+
+    def get_full_config(self):
+        ans = []
+        config_lines = self.generate_pgru_config()
+
+        for line in config_lines:
+            for config_name in ['ref', 'final']:
+                # we do not support user specified matrices in LSTM initialization
+                # so 'ref' and 'final' configs are the same.
+                ans.append((config_name, line))
+        return ans
+
+    # convenience function to generate the PGRU config
+    def generate_pgru_config(self):
+
+        # assign some variables to reduce verbosity
+        name = self.name
+        # in the below code we will just call descriptor_strings as descriptors for conciseness
+        input_dim = self.descriptors['input']['dim']
+        input_descriptor = self.descriptors['input']['final-string']
+        cell_dim = self.config['cell-dim']
+        rec_proj_dim = self.config['recurrent-projection-dim']
+        nonrec_proj_dim = self.config['non-recurrent-projection-dim']
+        delay = self.config['delay']
+        repair_nonlin = self.config['self-repair-scale-nonlinearity']
+        repair_nonlin_str = "self-repair-scale={0:.10f}".format(repair_nonlin) if repair_nonlin is not None else ''
+        bptrunc_str = ("clipping-threshold={0}"
+                      " zeroing-threshold={1}"
+                      " zeroing-interval={2}"
+                      " recurrence-interval={3}"
+                      "".format(self.config['clipping-threshold'],
+                                self.config['zeroing-threshold'],
+                                self.config['zeroing-interval'],
+                                abs(delay)))
+        affine_str = self.config['ng-affine-options']
+        pes_str = self.config['ng-per-element-scale-options']
+
+        # Natural gradient per element scale parameters
+        # TODO: decide if we want to keep exposing these options
+        if re.search('param-mean', pes_str) is None and \
+           re.search('param-stddev', pes_str) is None:
+           pes_str += " param-mean=0.0 param-stddev=1.0 "
+
+        # string for GruNonlinearityComponent
+        gru_nonlin_str = self.config['gru-nonlinearity-options']
+        
+        # formulation like:
+        # z_t = \sigmoid ( U^z x_t + W^z s_{t-1} )   # update gate
+        # r_t = \sigmoid ( U^r x_t + W^r s_{t-1} )   # reset gate
+        # h_t = \tanh ( U^h x_t + W^h ( s_{t-1} \dot r_t ) )
+        # c_t = ( 1 - z_t ) \dot h_t  +  z_t \dot c_{t-1}
+        # y_t = W^y c_t  # dim(y_t) = recurrent_dim + non_recurrent_dim.
+                         #  This is the output of the GRU.
+        # s_t = y_t[0:recurrent_dim-1]  # dimension range of y_t 
+                                        # dim(s_t) = recurrent_dim.
```

Comment:
I found c_t is used. I recommend the style \tilde{h}_t. Keep the same with the existing style, also keep the same with paper: https://arxiv.org/pdf/1803.10225.pdf & http://www.danielpovey.com/files/2018_interspeech_opgru.pdf

same goes for fast-(norm-)(O)pgru
```
        # formulation like:
        # z_t = \sigmoid ( x_t * U^z + s_{t-1} * W^z ) // update gate
        # r_t = \sigmoid ( x_t * U^r + s_{t-1} * W^r ) // reset gate
        # \tilde{h}_t = \tanh ( x_t * U^h + ( s_{t-1} \dot r_t ) * W^h )
        # h_t = ( 1 - z_t ) \dot \tilde{h}_t + z_t \dot h_{t-1}
        # y_t = h_t * W^y
        # s_t = y_t (0:rec_proj_dim-1)
```

Reply:
I saw Dan use "c_t" in GruNonlinearityComponent annotation and code. So I use it here so that the reader can understand the code easily.